Terminal.Gui/Views/TextMaskField.cs

```diff
@@ -0,0 +1,535 @@
+					if (cursorPosition > 1) {
+						text.RemoveAt (--cursorPosition);
+						TextChanged?.Invoke (Text);
+					}
+				} else {
+					if (cursorPosition > 0) {
+						text.RemoveAt (--cursorPosition);
+						TextChanged?.Invoke (Text);
+					}
+				}
+				break;
+			}
+		}
+
+		void DeleteKeyHandler ()
+		{
+			switch (masktype) {
+			case TextMaskType.Mask:
+				text [cursorPosition] = EmptyRune;
+				TextChanged?.Invoke (Text);
+				break;
+			case TextMaskType.Regex:
+				if (TextAlignment == TextAlignment.Right) {
+					if (cursorPosition > 0) {
+						text.RemoveAt (--cursorPosition);
+						TextChanged?.Invoke (Text);
+					}
+				} else {
+					if (text.Count > 0 && cursorPosition < text.Count) {
+						text.RemoveAt (cursorPosition);
+						TextChanged?.Invoke (Text);
+					}
+				}
+				break;
+			}
+		}
+
+		void HomeKeyHandler ()
+		{
+			switch (masktype) {
+			case TextMaskType.Mask:
+				cursorPosition = -1;
+				IncCursorPosition ();
+				break;
+			case TextMaskType.Regex:
+				cursorPosition = 0;
+				break;
+			}
+		}
+
+		void EndKeyHandler ()
+		{
+			switch (masktype) {
+			case TextMaskType.Mask:
+				cursorPosition = text.Count;
+				DecCursorPosition ();
+				break;
+			case TextMaskType.Regex:
+				cursorPosition = text.Count;
+				break;
+			}
+		}
+
+		///<inheritdoc/>
+		public override bool ProcessKey (KeyEvent kb)
+		{
+			switch (kb.Key) {
+			case Key.Home: HomeKeyHandler (); break;
+			case Key.End: EndKeyHandler (); break;
+			case Key.Delete:
+			case Key.DeleteChar: DeleteKeyHandler (); break;
+			case Key.Backspace: BackspaceKeyHandler (); break;
+			case Key.CursorLeft: DecCursorPosition (); break;
+			case Key.CursorRight: IncCursorPosition (); break;
+			default:
+				if (kb.Key < Key.Space || kb.Key > Key.CharMask)
+					return false;
+
+				var key = new Rune ((uint)kb.KeyValue);
+
+				switch (masktype) {
+				case TextMaskType.Mask:
+					if (ValidateRune (cursorPosition, key)) {
+						text [cursorPosition] = key;
+						TextChanged?.Invoke (Text);
+						IncCursorPosition ();
+					}
+					break;
+				case TextMaskType.Regex:
+					// Validate text copy first.
+					var aux = text.ToList ();
+					aux.Insert (cursorPosition, key);
+					if (Validate (aux) || RegexValidateInput == false) {
+						text.Insert (cursorPosition, key);
+						TextChanged?.Invoke (Text);
+						IncCursorPosition ();
+					}
+					break;
+				default:
+					throw new ArgumentOutOfRangeException ($"{masktype} not implemented.");
+				}
+				break;
+			}
+
+			// Used for style
+			error = Validate (text) ? false : true;
+
+			SetNeedsDisplay ();
+			return true;
+		}
+
+		/// <summary>
+		/// Valid mask characters and regex validator.
+		/// </summary>
+		protected Dictionary<Rune, string> MaskRunes { get; set; } = new Dictionary<Rune, string> {
+			['0'] = "[0-9]",
+			['9'] = "[0-9\\s]?",
+			['L'] = "[a-zA-Z]",
+			['?'] = "[a-zA-Z\\s]?",
+			['A'] = "[0-9a-zA-Z]",
+			['a'] = "[0-9a-zA-Z\\s]?"
+		};
```

Comment:
What if I wanted a field that only accepted non-alpha and non-numeric characters? e.g. `@#$%^&*`.

I suppose I could use a regex?